### PR TITLE
More readable message in ArrayIndexError

### PR DESCRIPTION
**Pull request by nordlow**
FYI, @andralex.

### CODE REVIEW

**Review by maxhaton — APPROVED** [review]

**Comment by CyberShadow** [comment]
Is the intention to fix the wording when the index equals the length?

**Comment by nordlow** [comment]
> Is the intention to fix the wording when the index equals the length?

This change on behalf of @andralex.

**Comment by CyberShadow** [comment]
Not to bikeshed but "too large" makes me think of an integer value that doesn't fit in a `size_t`.

Can we look at what terminology / phrasing other languages do, and stick to that? Maybe just "out of bounds"?

**Comment by andralex** [comment]
Good point. How about 

Index [7] is beyond the end of array of length 3.

Native speakers please opine!

**Comment by dkorpel** [comment]
> This change is on behalf of @andralex so safest to let him answer this question aswell.

Where did this come up?

> How about
>
> Index [7] is beyond the end of array of length 3.

That sounds equivalent to 'exceeds', which means to "go beyond what is allowed or stipulated by (a set limit)" or "go past an allowed limit".

I'm not against changing the message, I'm just curious what's unclear about it. After I wrote the new array error messages I asked a programmer who doesn't program in D whether it was clear and he thought so. Is 'exceed' a word people don't know or don't expect in this context?

**Comment by andralex** [comment]
It's not unclear, it's grammatically incorrect.

An index does not exceed an array. It exceeds the length of the array. The meaning of the message is that whatever index you give must be less than the length.

**Comment by andralex** [comment]
How about this - no repetitions, no akwardness:

`Index 7 out of bounds for array of length 3.`

Worx?

**Comment by dkorpel** [comment]
> An index does not exceed an array. It exceeds the length of the array. 

Ah, makes sense. I originally wrote "index [20] exceeds array length 20" but I added the 'of' because it looks weird that 20 exceeds 20.

> Index 7 out of bounds for array of length 3.

Sounds good

**Comment by andralex** [comment]
@dkorpel thanks, and also thank you for putting in the actual work!

**Comment by moon-chilled** [comment]
SBCL uses: 'Invalid index 7, should be a non-negative integer below 3'.  I like that because it's clear (esp. for newbies) in the case when the index is the array length.

**Comment by andralex** [comment]
love it!

**Comment by dkorpel** [comment]
'non-negative' is redundant in D's case, since array indices have the unsigned `size_t` type.

**Comment by thewilsonator** [comment]
> Index 7 out of bounds for array of length 3.

This leads to `Index 3 out of bounds for array of length 3` for an off-by-one, potentially quite confusing for a newb.

How about "Index 7 is out of bounds of range [0 .. 2]"?

**Comment by dkorpel** [comment]
> How about "Index 7 is out of bounds of range [0 .. 2]"?

Then `arr[0..3][3]` would result in "Index 3 is out of bounds of range [0 .. 2]"

**Comment by nordlow** [comment]
I'm gonna wait and see what fighter that still stands after a couple of more rounds... ;)

**Comment by RazvanN7** [comment]
"Index %s is out of bounds for array of length %s" is fine with me.

**Review by dkorpel — CHANGES_REQUESTED** [review]
Also update the comment on line 100:

> It's essentially `printf("index [%zu] exceeds array of length [%zu]", index, length)`

**Comment by nordlow** [comment]
> Also update the comment on line 100:
> 
> > It's essentially `printf("index [%zu] exceeds array of length [%zu]", index, length)`

Done.

**Review by RazvanN7 — APPROVED** [review]

File: src/core/exception.d [hunk 1]
```diff
@@ -107,7 +107,7 @@ class ArrayIndexError : RangeError
         sink.rangeMsgPut("index [");
         sink.rangeMsgPut(unsignedToTempString!10(index, tmpBuf));
         sink.rangeMsgPut("]");
-        sink.rangeMsgPut(" exceeds array of length ");
+        sink.rangeMsgPut(" is out of bounds for array of length ");
```
Comment (andralex) [inline src/core/exception.d:110]: Thanks @nordlow ! Not about this PR, but seems like like 109 and 110 could be merged.

Comment (nordlow) [reply src/core/exception.d:110]: Done at https://github.com/dlang/druntime/pull/3575.

**Pull request merged**